### PR TITLE
GSL: New version 2.7.1

### PR DESCRIPTION
**Pull request by eschnett**
Do we need to increase the "official" version number to 2.7.2, or should we keep it at 2.7.1?

### CODE REVIEW

**Comment by giordano** [comment]
I'd keep it to 2.7.1 :slightly_smiling_face:

**Comment by eschnett** [comment]
This fails for the architecture `aarch64-apple-darwin`.

I investigated, and I don't understand why. Apparently the symbols in the `libgslcblas` library aren't found when `libgsl` is created. The library is there, and it has the correct symbols. Maybe Apple's linker behaves differently on `aarch64` than on `x86_64`, and one needs to pass an additional option when creating `libgsl`. Ideally, this would happen via autotools and libtool, of course. Maybe updating these and running `autoreconf` would help?

This architecture isn't currently supported either. Would it be okay to disable it?

**Comment by giordano** [comment]
Yes, I agree that ideally autotools should deal with this stuff.  The weird thing is that v2.7.0 worked flawless.  I'll try to have a look these days.

**Comment by eschnett** [comment]
I didn't see a `GSL_jll @2.7` for `aarch64-apple-darwin` either. Did I miss it?

**Comment by giordano** [comment]
That's the "fake" v2.7.1: https://github.com/JuliaBinaryWrappers/GSL_jll.jl/blob/847c713027ebb947f26f2f140bd7cb86d222d2f2/README.md.  It was actually 2.7.0, see the source.

**Comment by eschnett** [comment]
Yes, I see it know. I accidentally looked at `GSL @2.6.0+3`.

**Comment by giordano** [comment]
Quick update: the [failing command](https://dev.azure.com/JuliaPackaging/Yggdrasil/_build/results?buildId=20104&view=logs&j=2ea53ff3-6db4-5249-2f87-da31e5d82122&t=67e8a9f6-14e3-5f6a-4b95-09ca0e7ad8bd&l=5590) is
```
libtool: link: cc -dynamiclib  -o .libs/libgsl.27.dylib  .libs/version.o   -Wl,-force_load,block/.libs/libgslblock.a -Wl,-force_load,blas/.libs/libgslblas.a -Wl,-force_load,bspline/.libs/libgslbspline.a -Wl,-force_load,bst/.libs/libgslbst.a -Wl,-force_load,complex/.libs/libgslcomplex.a -Wl,-force_load,cheb/.libs/libgslcheb.a -Wl,-force_load,dht/.libs/libgsldht.a -Wl,-force_load,diff/.libs/libgsldiff.a -Wl,-force_load,deriv/.libs/libgslderiv.a -Wl,-force_load,eigen/.libs/libgsleigen.a -Wl,-force_load,err/.libs/libgslerr.a -Wl,-force_load,fft/.libs/libgslfft.a -Wl,-force_load,filter/.libs/libgslfilter.a -Wl,-force_load,fit/.libs/libgslfit.a -Wl,-force_load,histogram/.libs/libgslhistogram.a -Wl,-force_load,ieee-utils/.libs/libgslieeeutils.a -Wl,-force_load,integration/.libs/libgslintegration.a -Wl,-force_load,interpolation/.libs/libgslinterpolation.a -Wl,-force_load,linalg/.libs/libgsllinalg.a -Wl,-force_load,matrix/.libs/libgslmatrix.a -Wl,-force_load,min/.libs/libgslmin.a -Wl,-force_load,monte/.libs/libgslmonte.a -Wl,-force_load,multifit/.libs/libgslmultifit.a -Wl,-force_load,multifit_nlinear/.libs/libgslmultifit_nlinear.a -Wl,-force_load,multilarge/.libs/libgslmultilarge.a -Wl,-force_load,multilarge_nlinear/.libs/libgslmultilarge_nlinear.a -Wl,-force_load,multimin/.libs/libgslmultimin.a -Wl,-force_load,multiroots/.libs/libgslmultiroots.a -Wl,-force_load,ntuple/.libs/libgslntuple.a -Wl,-force_load,ode-initval/.libs/libgslodeiv.a -Wl,-force_load,ode-initval2/.libs/libgslodeiv2.a -Wl,-force_load,permutation/.libs/libgslpermutation.a -Wl,-force_load,combination/.libs/libgslcombination.a -Wl,-force_load,multiset/.libs/libgslmultiset.a -Wl,-force_load,poly/.libs/libgslpoly.a -Wl,-force_load,qrng/.libs/libgslqrng.a -Wl,-force_load,randist/.libs/libgslrandist.a -Wl,-force_load,rng/.libs/libgslrng.a -Wl,-force_load,roots/.libs/libgslroots.a -Wl,-force_load,siman/.libs/libgslsiman.a -Wl,-force_load,sort/.libs/libgslsort.a -Wl,-force_load,specfunc/.libs/libgslspecfunc.a -Wl,-force_load,movstat/.libs/libgslmovstat.a -Wl,-force_load,rstat/.libs/libgslrstat.a -Wl,-force_load,statistics/.libs/libgslstatistics.a -Wl,-force_load,sum/.libs/libgslsum.a -Wl,-force_load,sys/.libs/libgslsys.a -Wl,-force_load,test/.libs/libgsltest.a -Wl,-force_load,utils/.libs/libutils.a -Wl,-force_load,vector/.libs/libgslvector.a -Wl,-force_load,cdf/.libs/libgslcdf.a -Wl,-force_load,wavelet/.libs/libgslwavelet.a -Wl,-force_load,spmatrix/.libs/libgslspmatrix.a -Wl,-force_load,spblas/.libs/libgslspblas.a -Wl,-force_load,splinalg/.libs/libgslsplinalg.a  -lm  -g -O2   -install_name  /workspace/destdir/lib/libgsl.27.dylib -compatibility_version 28 -current_version 28.0 -Wl,-single_module
Undefined symbols for architecture arm64:
  "_cblas_caxpy", referenced from:
      _gsl_blas_caxpy in libgslblas.a(blas.o)
  "_cblas_ccopy", referenced from:
      _gsl_blas_ccopy in libgslblas.a(blas.o)
    [...]
```
Instead, on x86_64-darwin [the compiler call](https://dev.azure.com/JuliaPackaging/Yggdrasil/_build/results?buildId=20104&view=logs&j=4ad5501b-6adf-545d-1cbf-e5f6d4fedafb&t=325a2760-47a9-5bf0-821f-6cf85ac7400a&l=5608) is
```
cc -dynamiclib -Wl,-undefined -Wl,dynamic_lookup -o .libs/libgsl.27.dylib  .libs/version.o   -Wl,-force_load,block/.libs/libgslblock.a -Wl,-force_load,blas/.libs/libgslblas.a -Wl,-force_load,bspline/.libs/libgslbspline.a -Wl,-force_load,bst/.libs/libgslbst.a -Wl,-force_load,complex/.libs/libgslcomplex.a -Wl,-force_load,cheb/.libs/libgslcheb.a -Wl,-force_load,dht/.libs/libgsldht.a -Wl,-force_load,diff/.libs/libgsldiff.a -Wl,-force_load,deriv/.libs/libgslderiv.a -Wl,-force_load,eigen/.libs/libgsleigen.a -Wl,-force_load,err/.libs/libgslerr.a -Wl,-force_load,fft/.libs/libgslfft.a -Wl,-force_load,filter/.libs/libgslfilter.a -Wl,-force_load,fit/.libs/libgslfit.a -Wl,-force_load,histogram/.libs/libgslhistogram.a -Wl,-force_load,ieee-utils/.libs/libgslieeeutils.a -Wl,-force_load,integration/.libs/libgslintegration.a -Wl,-force_load,interpolation/.libs/libgslinterpolation.a -Wl,-force_load,linalg/.libs/libgsllinalg.a -Wl,-force_load,matrix/.libs/libgslmatrix.a -Wl,-force_load,min/.libs/libgslmin.a -Wl,-force_load,monte/.libs/libgslmonte.a -Wl,-force_load,multifit/.libs/libgslmultifit.a -Wl,-force_load,multifit_nlinear/.libs/libgslmultifit_nlinear.a -Wl,-force_load,multilarge/.libs/libgslmultilarge.a -Wl,-force_load,multilarge_nlinear/.libs/libgslmultilarge_nlinear.a -Wl,-force_load,multimin/.libs/libgslmultimin.a -Wl,-force_load,multiroots/.libs/libgslmultiroots.a -Wl,-force_load,ntuple/.libs/libgslntuple.a -Wl,-force_load,ode-initval/.libs/libgslodeiv.a -Wl,-force_load,ode-initval2/.libs/libgslodeiv2.a -Wl,-force_load,permutation/.libs/libgslpermutation.a -Wl,-force_load,combination/.libs/libgslcombination.a -Wl,-force_load,multiset/.libs/libgslmultiset.a -Wl,-force_load,poly/.libs/libgslpoly.a -Wl,-force_load,qrng/.libs/libgslqrng.a -Wl,-force_load,randist/.libs/libgslrandist.a -Wl,-force_load,rng/.libs/libgslrng.a -Wl,-force_load,roots/.libs/libgslroots.a -Wl,-force_load,siman/.libs/libgslsiman.a -Wl,-force_load,sort/.libs/libgslsort.a -Wl,-force_load,specfunc/.libs/libgslspecfunc.a -Wl,-force_load,movstat/.libs/libgslmovstat.a -Wl,-force_load,rstat/.libs/libgslrstat.a -Wl,-force_load,statistics/.libs/libgslstatistics.a -Wl,-force_load,sum/.libs/libgslsum.a -Wl,-force_load,sys/.libs/libgslsys.a -Wl,-force_load,test/.libs/libgsltest.a -Wl,-force_load,utils/.libs/libutils.a -Wl,-force_load,vector/.libs/libgslvector.a -Wl,-force_load,cdf/.libs/libgslcdf.a -Wl,-force_load,wavelet/.libs/libgslwavelet.a -Wl,-force_load,spmatrix/.libs/libgslspmatrix.a -Wl,-force_load,spblas/.libs/libgslspblas.a -Wl,-force_load,splinalg/.libs/libgslsplinalg.a  -lm  -g -O2   -install_name  /workspace/destdir/lib/libgsl.27.dylib -compatibility_version 28 -current_version 28.0 -Wl,-single_module
```
The only difference are the flags `-Wl,-undefined -Wl,dynamic_lookup` passed to the linker, which don't appear in the compiler call for aarch64-darwin.  I'll need to understand why they're called differently.

Side note: I've just noticed that the new soversion of the library is 27.  It used to be 25 in v2.7, that's an ABI breakage... At this point I fear it'd be better to keep the two versions separate :confused:

**Comment by giordano** [comment]
And the reason why we get the wrong are probably these lines in `aclocal.m4`:
```
    darwin*) # darwin 5.x on
      # if running on 10.5 or later, the deployment target defaults
      # to the OS version, if on x86, and 10.4, the deployment
      # target defaults to 10.4. Don't you love it?
      case ${MACOSX_DEPLOYMENT_TARGET-10.0},$host in
	10.0,*86*-darwin8*|10.0,*-darwin[[91]]*)
	  _lt_dar_allow_undefined='$wl-undefined ${wl}dynamic_lookup' ;;
	10.[[012]][[,.]]*)
	  _lt_dar_allow_undefined='$wl-flat_namespace $wl-undefined ${wl}suppress' ;;
	10.*)
	  _lt_dar_allow_undefined='$wl-undefined ${wl}dynamic_lookup' ;;
      esac
```
Needless to say that they just don't know there are new version of macOS after 10:
```console
sandbox:${WORKSPACE}/srcdir/gsl-2.7.1 # echo ${MACOSX_DEPLOYMENT_TARGET}
11.0
```
Sigh

File: G/GSL/GSL@2/build_tarballs.jl [hunk 1]
```diff
@@ -3,17 +3,26 @@
...
 
 # Bash recipe for building across all platforms
 script = raw"""
 cd $WORKSPACE/srcdir/gsl-*/
+if [[ "${target}" == aarch64-apple-darwin* ]]; then
+    # aclocal.m4 has some lines where it expects `MACOSX_DEPLOYMENT_TARGET` to be up to
+    # version 10.  Let's pretend to be 10.16, as many tools do to make old build systems
+    # happy.
+    export MACOSX_DEPLOYMENT_TARGET="10.16"
+fi
```
Comment (giordano) [inline G/GSL/GSL@2/build_tarballs.jl:25]: This also explains why the previous build worked: before https://github.com/JuliaPackaging/BinaryBuilderBase.jl/pull/217 we used to unconditionally set `MACOSX_DEPLOYMENT_TARGET="10.8"`.

**Pull request merged**